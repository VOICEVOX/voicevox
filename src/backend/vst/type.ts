import { TrackId } from "@/type/preload";

export type Routing = {
  channelMode: "mono" | "stereo";
  channelIndex: Record<TrackId, number>;
};
