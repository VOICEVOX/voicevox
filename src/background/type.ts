import { z } from "zod";
import { engineIdSchema } from "@/type/preload";

// デフォルトエンジン情報
export const defaultEngineInfosEnvSchema = z
  .object({
    uuid: engineIdSchema,
    host: z.string(),
    name: z.string(),
    executionEnabled: z.boolean(),
    executionFilePath: z.string(),
    executionArgs: z.array(z.string()),
    path: z.string().optional(),
  })
  .array();
