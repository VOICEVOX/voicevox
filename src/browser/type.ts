import { IpcIHData } from "@/type/ipc";

export type WorkerToMainMessage = {
  [K in keyof IpcIHData]: {
    type: K;
    return: IpcIHData[K]["return"];
    eventId: string;
  };
}[keyof IpcIHData];

export type MainToWorkerMessage = {
  [K in keyof IpcIHData]: {
    type: K;
    args: IpcIHData[K]["args"];
    eventId: string;
  };
}[keyof IpcIHData];
