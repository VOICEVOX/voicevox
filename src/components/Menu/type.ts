export type MenuItemBase<T extends string> = {
  type: T;
  label?: string;
};

export type MenuItemSeparator = MenuItemBase<"separator">;

export type MenuItemRoot = MenuItemBase<"root"> & {
  onClick?: () => void;
  subMenu: MenuItemData[];
  icon?: string;
  disabled?: boolean;
  disableWhenUiLocked: boolean;
  disablreloadingLocked?: boolean;
};

export type MenuItemButton = MenuItemBase<"button"> & {
  onClick: () => void;
  icon?: string;
  disabled?: boolean;
  disableWhenUiLocked: boolean;
  disablreloadingLocked?: boolean;
};

export type MenuItemData = MenuItemSeparator | MenuItemRoot | MenuItemButton;

export type MenuItemType = MenuItemData["type"];
