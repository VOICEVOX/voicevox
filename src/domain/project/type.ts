import { z } from "zod";

import {
  noteSchema,
  phonemeTimingEditSchema,
  singerSchema,
  tempoSchema,
  timeSignatureSchema,
  trackSchema,
} from "@/domain/project/schema";
import { NoteId } from "@/type/preload";

export type Tempo = z.infer<typeof tempoSchema>;

export type TimeSignature = z.infer<typeof timeSignatureSchema>;

export type Note = z.infer<typeof noteSchema>;

export type Singer = z.infer<typeof singerSchema>;

export type Track = z.infer<typeof trackSchema>;

export type PhonemeTimingEdit = z.infer<typeof phonemeTimingEditSchema>;

export type PhonemeTimingEditData = Map<NoteId, PhonemeTimingEdit[]>;
