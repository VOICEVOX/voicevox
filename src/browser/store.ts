import { defaultEngine, directoryHandleStoreKey } from "./contract";
import {
  EngineId,
  electronStoreSchema,
  engineSettingSchema,
} from "@/type/preload";

const dbName = "voicevox-web";
// FIXME: DBのschemaを変更したら、dbVersionを上げる
// TODO: 気づけるようにしたい
const dbVersion = 1;
// NOTE: settingを複数持つことはないと仮定して、keyを固定してしまう
export const entryKey = "value";

export const openDB = () =>
  new Promise<IDBDatabase>((resolve, reject) => {
    const request = indexedDB.open(dbName, dbVersion);
    request.onsuccess = () => {
      resolve(request.result);
    };
    request.onerror = () => {
      // TODO: handling
      reject(request.error);
    };
    request.onupgradeneeded = (ev) => {
      if (ev.oldVersion === 0) {
        // Initialize
        const db = request.result;
        const baseSchema = electronStoreSchema.parse({});
        Object.entries(baseSchema).forEach(([key, value]) => {
          const k = key as keyof typeof baseSchema;
          if (k !== "engineSettings") {
            db.createObjectStore(key).add(value, entryKey);
            return;
          }
          // defaultのEngineSettingを追加
          const defaultVoicevoxEngineId = EngineId(defaultEngine.uuid);
          db.createObjectStore(key).add(
            {
              [defaultVoicevoxEngineId]: engineSettingSchema.parse({}),
            },
            entryKey
          );
        });

        // DirectoryHandleも格納する
        db.createObjectStore(directoryHandleStoreKey);
      }
      // TODO: migrate
    };
  });
