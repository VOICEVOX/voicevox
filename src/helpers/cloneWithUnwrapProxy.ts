import createRfdc from "rfdc";

const rfdc = createRfdc();

/** Proxyを展開してクローンする。*/
export const cloneWithUnwrapProxy = <T>(obj: T): T => rfdc(obj);
