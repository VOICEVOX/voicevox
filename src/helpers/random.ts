/**
 * 乱数値を生成する。モックに対応している。
 * モックモードでは呼ばれた回数に応じて固定の値を返す。
 */

let mockMode = false;
let mockCount = 0;

/**
 * モックモードにし、呼ばれた回数をリセットする。
 */
export function resetMockMode(): void {
  mockMode = true;
  mockCount = 0;
}

/**
 * v4 UUID を生成する。
 */
export function uuid4(): string {
  if (!mockMode) {
    return crypto.randomUUID();
  } else {
    mockCount++;
    return `00000000-0000-4000-0000-${mockCount.toString().padStart(12, "0")}`;
  }
}
