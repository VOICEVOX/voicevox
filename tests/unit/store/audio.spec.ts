import { initializeStateAsSoftwareStarted, mockSandbox } from "./utils";

import { createStore } from "@/store";
import { resetMockMode } from "@/helpers/random";
import { cloneWithUnwrapProxy } from "@/helpers/cloneWithUnwrapProxy";
import { proxyStoreCreator } from "@/store/proxy";
import { UnreachableError } from "@/type/utility";
import { OpenAPIMockEngineConnectorFactory } from "@/infrastructures/EngineConnector";

const store = createStore({
  proxyStoreDI: proxyStoreCreator(OpenAPIMockEngineConnectorFactory),
});
const initialState = cloneWithUnwrapProxy(store.state);
beforeEach(() => {
  store.replaceState(initialState);

  resetMockMode();
  mockSandbox();
});

describe("GENERATE_AUDIO_ITEM", () => {
  beforeEach(async () => {
    await initializeStateAsSoftwareStarted(store);
  });

  it("何も指定しない場合は空のAudioItemを作る", async () => {
    const audioItem = await store.actions.GENERATE_AUDIO_ITEM({});
    expect(audioItem).toMatchSnapshot();
  });

  it("テキストを指定するとtextに値が入る", async () => {
    const audioItem = await store.actions.GENERATE_AUDIO_ITEM({
      text: "こんにちは",
    });
    expect(audioItem).toMatchSnapshot();
  });

  it("baseAudioItemを指定するとパラメータを引き継ぐ", async () => {
    const baseAudioItem = await store.actions.GENERATE_AUDIO_ITEM({});
    if (baseAudioItem.query == undefined) throw new UnreachableError();

    baseAudioItem.query.speedScale = 2.0;

    const audioItem = await store.actions.GENERATE_AUDIO_ITEM({
      baseAudioItem,
    });
    expect(audioItem.query?.speedScale).toBe(2.0);
  });
});
