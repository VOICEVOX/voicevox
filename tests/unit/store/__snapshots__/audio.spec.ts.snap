// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GENERATE_AUDIO_ITEM > テキストを指定するとtextに値が入る 1`] = `
{
  "presetKey": undefined,
  "query": {
    "accentPhrases": [
      {
        "accent": 5,
        "moras": [
          {
            "consonant": "k",
            "consonantLength": 0.0220625,
            "pitch": 4.7734375,
            "text": "コ",
            "vowel": "o",
            "vowelLength": 0.1140625,
          },
          {
            "consonant": undefined,
            "consonantLength": undefined,
            "pitch": 3.609375,
            "text": "ン",
            "vowel": "N",
            "vowelLength": 0.08312499999999999,
          },
          {
            "consonant": "n",
            "consonantLength": 0.022625,
            "pitch": 4.75,
            "text": "ニ",
            "vowel": "i",
            "vowelLength": 0.10843749999999999,
          },
          {
            "consonant": "ch",
            "consonantLength": 0.0400625,
            "pitch": 4.0234375,
            "text": "チ",
            "vowel": "i",
            "vowelLength": 0.10843749999999999,
          },
          {
            "consonant": "h",
            "consonantLength": 0.0315,
            "pitch": 4.796875,
            "text": "ハ",
            "vowel": "a",
            "vowelLength": 0.1509375,
          },
        ],
        "pauseMora": undefined,
      },
    ],
    "intonationScale": 1,
    "outputSamplingRate": 24000,
    "outputStereo": false,
    "pauseLengthScale": 1,
    "pitchScale": 0,
    "postPhonemeLength": 0.1,
    "prePhonemeLength": 0.1,
    "speedScale": 1,
    "volumeScale": 1,
  },
  "text": "こんにちは",
  "voice": {
    "engineId": "c7b58856-bd56-4aa1-afb7-b8415f824b06",
    "speakerId": "7ffcb7ce-00ec-4bdc-82cd-45a8889e43ff",
    "styleId": 0,
  },
}
`;

exports[`GENERATE_AUDIO_ITEM > 何も指定しない場合は空のAudioItemを作る 1`] = `
{
  "presetKey": undefined,
  "query": {
    "accentPhrases": [],
    "intonationScale": 1,
    "outputSamplingRate": 24000,
    "outputStereo": false,
    "pauseLengthScale": 1,
    "pitchScale": 0,
    "postPhonemeLength": 0.1,
    "prePhonemeLength": 0.1,
    "speedScale": 1,
    "volumeScale": 1,
  },
  "text": "",
  "voice": {
    "engineId": "c7b58856-bd56-4aa1-afb7-b8415f824b06",
    "speakerId": "7ffcb7ce-00ec-4bdc-82cd-45a8889e43ff",
    "styleId": 0,
  },
}
`;
